<!DOCTYPE html PUBLIC
  "-//W3C//DTD XHTML 1.1//EN"
  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">

<html>
<head>
    <title>GADGET User Manual</title>
    
    <link rel="shortcut icon" href="/static/favicon.png" />
    
    <link rel="stylesheet" type="text/css" href="/static/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="/static/bootstrap-theme.min.css" />
    <link rel="stylesheet" type="text/css" href="/static/base.css" />
    
    <link rel="stylesheet" type="text/css" href="static/base.css" />
</head>
<body><div id="manual">

    <br /><a href="/" style="position:absolute;right:0">&larr; Back to GADGET</a><br /><br />

    <a name="top"><h1>GADGET User Manual</h1></a>
    
    <p><a href="/">GADGET</a> is a tool that lets you search for genes and metabolites in the biomedical literature.</p>
    
    
    <ul style="margin-top:40px">
        <li><a href="#genesearch">Search for genes</a></li>
        <li><a href="#metabolitesearch">Search for metabolites</a></li>
        <br />
        <li><a href="#queryterms">Searching by keyword</a></li>
        <li><a href="#geneentry">Searching by gene symbol</a></li>
        <li><a href="#scores">Gene / metabolite scores</a></li>
        <li><a href="#species">Selecting a species</a></li>
        <li><a href="#usehomologs">Optional inclusion of homologs</a></li>
        <br />
        <li><a href="#data_sources">Data sources</a></li>
        <li><a href="#data_api">Data API - Using GADGET without a browser</a></li>
    </ul>
    
    
    <a class="topic" name="genesearch">Search for genes</a>
    <p>When you search for genes, GADGET first finds <a href="http://www.ncbi.nlm.nih.gov/pubmed/" target="_blank">PubMed</a> articles that match your search.  From the set of articles we've found, then GADGET compiles a list of genes that are overrepresented in the articles, and thus related to your topic.  (If you've entered a keyword query, GADGET checks to see if your keywords are present in the abstract for each article.  If you've entered gene symbols, GADGET checks several data sources to find articles that mention your genes.)</p>
    <p>For each gene in the search results, you can see the abstracts that link this gene to your search by clicking the "show abstracts" link.</p>
    <p>You can download gene search results as a CSV file.  Downloaded results will contain all of the genes that match your query, not only the top-scoring genes displayed on the page.</p>
    

    <a class="topic" name="metabolitesearch">Search for metabolites</a>
    <p>GADGET's metabolite search shows you a list of metabolites related to your query.  It works the same as the gene search, by first finding a set of articles related to your query, and then compiling a list of metabolites from the articles (see above for details.)</p>
    


    <a class="topic" name="queryterms">Searching by keyword</a>
    <p>GADGET can find genes and metabolites that match any set of keywords or keyphrases, by finding PubMed abstracts that contain the keywords and keyphrases in your query.</p>
    <br />
    <p>GADGET supports Boolean queries for more specific searches.  Terms can be separated with <span class="example">AND</span> and <span class="example">OR</span>.  An <span class="example">AND</span> query will match abstracts that contain both terms, while an <span class="example">OR</span> query will match abstracts that contain either term.  Terms separated by spaces are automatically treated as an <span class="example">AND</span> query.  You can also exclude terms from your query by using <span class="example">NOT</span> followed by the term.  You can use parentheses to make complex Boolean queries.</p>
    <p>Examples:</p>
    <div class="example">Match genes or interactions referenced in abstracts matching both "Cat" and "Dog":<br /><span>Cat AND Dog</span> or just <span>Cat Dog</span></div>
    <div class="example">Match genes or interactions referenced in abstracts matching "Cat" or "Dog", but not "Fish":<br /><span>Cat OR Dog NOT Fish</span></div>
    <div class="example">Match genes or interactions referenced in abstracts matching "Cat", or both "Dog" and "Fish":<br /><span>Cat OR (Dog AND Fish)</span></div>
    <br />
    <p>To do an exact match on a multi-word phrase, enclose the phrase in quotes.</p>
    <div class="example">Match abstracts containing the terms "Embryonic", "Stem", and "Cell":<br /><span>Embryonic stem cell</span></div>
    <div class="example">Match abstracts containing "Embryonic stem cell" all together:<br /><span>"Embryonic stem cell"</span></div>
    <br />
    <p>You can also restrict abstracts by the year they were published by entering <span class="example">year:</span> followed by the year.  You can specify ranges of years by using <span class="example">year:[<u>startyear</u> to <u>endyear</u>]</span>.  You can leave either <u>startyear</u> or <u>endyear</u> blank for an open-ended range.</p>
    <div class="example">Match abstracts published in 2011:<br /><span>year:2011</span></div>
    <div class="example">Match abstracts containing "apoptosis" published between 2005 and 2009 (inclusive):<br /><span>apoptosis year:[2005 to 2009]</span></div>
    <div class="example">Match abstracts containing "surfactant" published in or before 2007:<br /><span>surfactant year:[to 2007]</span></div>
    <div class="example">Match abstracts containing "chitin" published in or after 2009:<br /><span>chitin year:[2009 to]</span></div>
    <br />
    <p>GADGET accepts wildcards to match inexact terms.  Including a question mark (<span class="example">?</span>) in a term will match any single character, while an asterisk (<span class="example">*</span>) will match any number of arbitrary characters.  Note that using a wildcard character at the beginning of a term will significantly slow GADGET down.  You cannot use wildcard characters within quotation marks.</p>
    <div class="example">Match abstracts containing "text", "test", etc.:<br /><span>te?t</span></div>
    <div class="example">Match abstracts containing "BLAST", "blastula", "blastema", etc.:<br /><span>blast*</span></div>
    <br />
    <p>Query terms in GADGET are not case sensitive (although Boolean operators are case sensitive).</p>
    <p>You will often get better and faster results by using more specific terms in your searches.  Avoid using common terms like "gene" in your query; they don't usually add much value to the results, and they can slow down GADGET.</p>
    <p>GADGET uses the <a href="http://packages.python.org/Whoosh/querylang.html" target="_blank">Whoosh query language</a> for matching abstracts.  Not all features of the query language are applicable to GADGET.</p>



<a class="topic" name="geneentry">Searching via related genes</a>
    <p>You can enter a list of genes to search for related genes and metabolites.  GADGET will find articles that mention your queried genes, and show you a list of genes/metabolites highly represented in those abstracts.</p>
    <p>You can enter genes using their <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene" target="_blank">Entrez Gene</a> symbols or Gene ID's (eg. <span class="example">TAL1</span> or <span class="example">6886</span>.)</p>
    <p>You can select whether GADGET should use articles that mention <span class="example">any</span> (at least one) or <span class="example">all</span> of your genes.</p>
    <p>You can use boolean operators (<span class="example">AND</span>, <span class="example">OR</span>, and <span class="example">NOT</span> in caps) in your gene list.  These operators override your <span class="example">any</span> / <span class="example">all</span> choice.</p>  You can also use <span class="example">*</span> for wildcard matches.
    <div class="example">With the <span>any</span> option selected, match abstracts that mention either SHH or LBH (or both.)  With the </span>all</span> option selected, match abstracts that mention both SHH and LBH.<br /><span>SHH LBH</span></div>
    <div class="example">The same as the previous example, using Entrez Gene ID's instead of gene symbols:<br /><span>6469 81606</span></div>
    <div class="example">Match abstracts that mention PECAM1 and either KDR or TIE1 (or both):<br /><span>PECAM1 AND (KDR OR TIE1)</span></div>
    <div class="example">Match abstracts that don't mention TWIST1:<br /><span>NOT TWIST1</span></div>
    <div class="example">Match all of the gene symbols that start with HOX<br /><span>HOX*</span></div>
    <p>You can also upload a file with a list of genes.  The file needs to have one gene symbol or Entrez Gene ID per line.</p>



<a class="topic" name="scores">Gene / metabolite scores</a>

    <p>GADGET can optionally use a few different scores to rank search results for genes and metabolites:</p>
    <table class="instructions">
        <tr><td>F1 Balanced Score</td><td>The <b>default</b> <a href="http://en.wikipedia.org/wiki/F1_score" target="_blank">F1 score</a> is a balanced measurement of <a href="http://en.wikipedia.org/wiki/Precision_and_recall" target="_blank">precision and recall</a> (specificity and query-matching abstracts).  It takes both the proportion of matching abstracts for each gene and the total number of matching abstracts into account.  The F1 score represents a tradeoff between the "Adjusted precision" and "Matching abstracts" scores.</td></tr>
        <tr><td>Specificity</td><td>The specificity score is based upon the proportion of a gene's abstracts that match your query &ndash; it hilights genes that occur frequently in articles matching your query, but occuring less frequently in other articles.  For each gene in your query, specificity is calculated as (number of abstracts for the gene matching the query) &divide; (10 + total number of abstracts for the gene).  The specificity score is adjusted by adding a "pseudocount" of 10 to the denominator, so that a gene with 10/10 matching abstracts scores higher than a gene with 1/1 matching abstracts.  (See <a href="http://en.wikipedia.org/wiki/Precision_%28information_retrieval%29" target="_blank">precision</a> on Wikipedia.)</td></tr>
        <tr><td>Query-matching abstracts</td><td>The number of abstracts for each gene/metabolite that match your query.</td></tr>
        <tr><td>Total abstracts</td><td>The total number of abstracts associated with each gene.  (This number is not based upon your query.)</td></tr>
    </table>
    <br />
    <p>GADGET also computes a <b>p-value</b> for each gene using the <a href="http://en.wikipedia.org/wiki/Hypergeometric_test" target="_blank">hypergeometric test</a>.  The null hypothesis for the test is that the fraction of abstracts that match the query for a given gene is equal to the "background" fraction of abstracts that match the query in the entire corpus of abstracts.  The p-value thus indicates how confident we are that a gene's abstract set is enriched for the query.  (There is no option in GADGET to sort genes by the p-value because it would be too slow, but you can still do this if you download your search results in a CSV file.)</p>



    <a class="topic" name="species">Selecting a species</a>

    <p><b>Gene search only</b></p>
    <p>When you do a GADGET <i>gene search</i>, all of the genes in your results will be from a particular species.  GADGET allows you to select from several species to view genes for.  The currently-supported species are:
    <table style="border-collapse:separate;border-spacing:15px;">
        <tr><th>NCBI Taxonomy ID</th><th>Name</th></tr>
        <tr><td>9606</td><td>Homo Sapiens (default)</td></tr>
        <tr><td>10090</td><td>Mus musculus</td></tr>
        <tr><td>559292</td><td>Saccharomyces cerevisiae</td></tr>
    </table></p>
    <p><b>The metabolite search currently only supports Homo Sapiens.</b></p>
    <p>For gene search queries, when you provide a list of <a href="#geneentry">gene symbols</a>, GADGET will match them to genes from your selected species.</p>




    <a class="topic" name="usehomologs">Optional inclusion of homologs</a>
    <p><b>Gene search only</b></p>
    <p>When finding and <a href="#scores">ranking</a> genes that match your query, GADGET's gene search can optionally take homologs into account.  To calculate scores for each gene, GADGET counts the number of abstracts matching your query that each gene appears in.  If you choose to include homologs, GADGET will also count abstracts that refer to homologs of each gene.</p>
    <p>The groups of homologs that GADGET uses come from <a href="http://www.ncbi.nlm.nih.gov/homologene" target="_blank">Homologene</a>.  Homologs can be either from the same species as each gene, or from different species.</p>
    <p>This feature can be useful for queries that match a small number of articles.</p>

    
    
    <a class="topic" name="data_sources">Data sources</a>
    <p>GADGET's gene search uses gene-article links from NCBI's "gene2pubmed" dataset, Saccharomyces Genome Database (SGD), and Mouse Genome Informatics (MGI).  Gene homologs come from the Homologene database.</p>
    <p>GADGET's metabolite search uses a list of metabolites from the Human Metabolome Database (HMDB).  Metabolite-article links are generated by matching metabolite names and synonyms from HMDB in the set of PubMed abstracts from the gene search.</p>
    <p>GADGET downloads abstract text from PubMed.  New abstracts, gene-abstract links, and metabolite-abstract links are updated regularly.</p>
    
    

    <a class="topic" name="data_api">Data API - Using GADGET without a browser</a>
    <p>If you want to automatically access GADGET without using a web browser, you can download comma-delimited CSV or XML files by fetching URL's from GADGET and sending arguments via the query string at the end of the URL.  To download data, the query string must include <span class="example">&download=CSV</span>.</p>
    <p>To search genes, use this URL: <span class="example">http://gadget.biostat.org/gadget/genelist?download=csv&<span style="font-weight:normal">&lt;query parameters here&gt;</span></span></p>
    <p>To search metabolites, use this URL: <span class="example">http://gadget.biostat.org/gadget/metabolitelist?download=csv&<span style="font-weight:normal">&lt;query parameters here&gt;</span></span></p>
    <p>To send information to GADGET, just append it to the url.  Aguments are given by the name of the argument, an equals sign (<span class="example">=</span>), and the argument's value, e.g. <span class="example"><u>argument</u>=<u>value</u></span>.  Multiple arguments should be separated by ampersands (<span class="example">&</span>).  Here's a list of the arguments you can send:</p>
    <table class="instructions">
        <tr><td>download</td><td>This argument is <b>required</b> to download the results as a CSV file. (<span>&download=CSV</span>.)</td></tr>
        <tr><td>q</td><td>The keyword query terms of your search.<br /><br />In most browsers, you can include spaces and quotation marks in the URL.  Special characters, like "&" and "/", must be <a href="http://en.wikipedia.org/wiki/Percent-encoding#Percent-encoding_reserved_characters" target="_blank">percent encoded</a>.<br /><br />Either <span>q</span> or <span>genes</span> (or both) is required for GADGET queries.</td></tr>
        <tr><td>genes</td><td>A list of gene symbols and/or Entrez Gene IDs to search for.  You can use the same format and all of the same <a href="#geneentry">features for entering genes</a> as you can in the graphical interface.<br /><br />If you provide gene symbols, GADGET will match them to genes for your selected <span>species</span><br /><br />Either <span>q</span> or <span>genes</span> (or both) is required for GADGET queries.</td></tr>
        <tr><td>geneop</td><td>The "gene operator" to use for your list of genes - whether to match abstracts that mention <span>any</span> or <span>all</span> of your genes (see <a href="#geneentry">gene entry for the gene search and keyphrase search</a>.)  If <span>geneop</span> isn't provided, <span>any</span> will be used by default.</td></tr>
        <tr><td>species</td><td><b>Gene search only</b> (GADGET's metabolite search only supports homo sapiens for now.)  <br /><br />The species to use for your gene search.  All of the genes in your gene search results will belong to this species.<br /><br />To specify a species, use the species's NCBI taxonomy ID, (e.g. <span>&species=9606</span> for humans).  See <a href="#species">here</a> for a list of currently available species and taxonomy IDs.  If you don't specify a species, <b>homo sapiens will be used by default.</b></td></tr>
        <tr><td>usehomologs</td><td><b>Gene search only</b><br /><br />Whether or not to include homologs when calculating scores for each gene.  (See <a href="#usehomologs">here</a>.)  To include homologs, include <span>&usehomologs=1</span> in your query string.  If <span>usehomologs</span> is omitted, homologs will not be included by default.</td></tr>
        <tr><td>orderby</td><td>Which score to use to rank the results.<br /><br />There are 4 possible values for <span>orderby</span>.  The possible values are <span>f1_score</span>, which will order the results by the F1 score; <span>adjusted_precision</span>, which will order the results by the specificity (aka. adjusted precision); <span>matching_abstracts</span>, which will order the results by the number of abstracts matching your query; and <span>total_abstracts</span>, which will order the results by the total number of abstracts for each gene/metabolite.<br /><br />If the <span>orderby</span> argument is omitted or invalid, GADGET will default to the F1 score.  See <a href="#scores">above</a> for more information about the different scores.</td></tr>
         <tr><td>limit</td><td>The maximum number of results to return.  This argument should be a number.  If <span>limit</span> is not included or is invalid, GADGET will return all of the genes/metabolites found in the search.</td></tr>
         <tr><td>offset</td><td>The rank (minus one) of the first gene/metabolite to include in the results.  To omit the first 100 results, include <span>&offset=100</span>.  If <span>offset</span> is omitted or invalid, GADGET will not omit any results.</td></tr>
    </table>
    <p>If no articles or genes/metabolites match your query, or if <span class="example">q</span> is omitted or invalid, the GADGET server will issue a 404 error instead of returning a CSV file.</p>
    <p>Examples:</p>
    <div class="example">Download genes matching "arteriole", using homologs:<br /><span>http://gadget.biostat.wisc.edu/gadget/genelist?download=CSV&usehomologs=1&q=arteriole</b></span></div>
    <div class="example">Download mouse genes matching "mesoderm", ranked by adjusted precision:<br /><span>http://gadget.biostat.wisc.edu/gadget/genelist?download=CSV&q=mesoderm&orderby=adjusted_precision</span></div>
    <div class="example">Download the top 150 genes for mice matching "Loop of Henle", using homologs:<br /><span>http://gadget.biostat.wisc.edu/gadget/genelist?download=CSV&species=10090&q=Loop of Henle&limit=150&usehomologs=1</span></div>
    <div class="example">Download the top 80 metabolites related to CFTR matching "cystic fibrosis", ranked by number of matching abstracts:<br /><span>http://gadget.biostat.wisc.edu/gadget/metabolitelist?download=CSV&q=cystic fibrosis&genes=CFTR&limit=80&orderby=matching_abstracts</span></div>

        



    
</div></body>
</html>
